### PR TITLE
Optimize Surface Modifiers

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
Changed the SurfaceData ModifySurfacePoints API again to be a bit more straightforward than before - the previous iteration had a strange call pattern back and forth from the component to the SurfacePointList back into the component. Now it just goes SurfacePointList to component, without bouncing back and forth. Also, by passing in the spans of data, the GradientSurfaceDataComponent is able to directly pass the positions span through to the bulk gradient API, allowing for even more speedups.

These are the components that have been optimized:
* GradientSurfaceDataComponent
* SurfaceDataShapeComponent
* SurfaceDataColliderComponent

Also added benchmarks for GradientSurfaceDataComponent to test the changes. Note that the single-query case has become 40% slower, but the multi-query case is > 50% faster. The single-query case is generally getting phased out of usage though, so the benefits in the multi-query case seem worth it. Regardless, I'll do a bit more research and potentially post a follow-up PR if I find a clean way to optimize for both cases.

Before:
```
-------------------------------------------------------------------------------------------------
Benchmark                                                       Time             CPU   Iterations
-------------------------------------------------------------------------------------------------
GradientSurfaceData/BM_GetSurfacePoints/1024                  390 ms          391 ms            2
GradientSurfaceData/BM_GetSurfacePoints/2048                 1578 ms         1578 ms            1
GradientSurfaceData/BM_GetSurfacePointsFromRegion/1024        138 ms          141 ms            4
GradientSurfaceData/BM_GetSurfacePointsFromRegion/2048        723 ms          719 ms            1
GradientSurfaceData/BM_GetSurfacePointsFromList/1024          140 ms          141 ms            5
GradientSurfaceData/BM_GetSurfacePointsFromList/2048          553 ms          547 ms            1
```

After:
```
-------------------------------------------------------------------------------------------------
Benchmark                                                       Time             CPU   Iterations
-------------------------------------------------------------------------------------------------
GradientSurfaceData/BM_GetSurfacePoints/1024                  517 ms          516 ms            1
GradientSurfaceData/BM_GetSurfacePoints/2048                 2071 ms         2063 ms            1
GradientSurfaceData/BM_GetSurfacePointsFromRegion/1024       67.0 ms         67.2 ms           10
GradientSurfaceData/BM_GetSurfacePointsFromRegion/2048        270 ms          273 ms            2
GradientSurfaceData/BM_GetSurfacePointsFromList/1024         65.6 ms         65.3 ms           11
GradientSurfaceData/BM_GetSurfacePointsFromList/2048          264 ms          258 ms            2
```

Before/after benchmarks for the generic SurfaceData benchmarks, which use the SurfaceDataShapeComponent, which already had most but not all of the optimizations:

Before:
```
------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------
SurfaceDataBenchmark/BM_GetSurfacePoints/1024                      673 ms          672 ms            1
SurfaceDataBenchmark/BM_GetSurfacePoints/2048                     2672 ms         2672 ms            1
SurfaceDataBenchmark/BM_GetSurfacePointsFromRegion/1024            207 ms          208 ms            3
SurfaceDataBenchmark/BM_GetSurfacePointsFromRegion/2048           1217 ms         1219 ms            1
SurfaceDataBenchmark/BM_GetSurfacePointsFromList/1024              212 ms          208 ms            3
SurfaceDataBenchmark/BM_GetSurfacePointsFromList/2048              819 ms          813 ms            1
```

After:
```
------------------------------------------------------------------------------------------------------
Benchmark                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------
SurfaceDataBenchmark/BM_GetSurfacePoints/1024                      686 ms          688 ms            1
SurfaceDataBenchmark/BM_GetSurfacePoints/2048                     2735 ms         2734 ms            1
SurfaceDataBenchmark/BM_GetSurfacePointsFromRegion/1024            189 ms          191 ms            4
SurfaceDataBenchmark/BM_GetSurfacePointsFromRegion/2048           1161 ms         1156 ms            1
SurfaceDataBenchmark/BM_GetSurfacePointsFromList/1024              185 ms          184 ms            4
SurfaceDataBenchmark/BM_GetSurfacePointsFromList/2048              742 ms          750 ms            1
```

### CODE REVIEW

**Review by cgalvan — APPROVED** [review]

File: Gems/GradientSignal/Code/Tests/GradientSignalBenchmarks.cpp [hunk 1]
```diff
@@ -160,6 +168,195 @@ namespace UnitTest
...
+            float halfWorldSize = worldSize / 2.0f;
+
+            // Create a large flat box with 2 provider tags.
+            {
+                AZStd::unique_ptr<AZ::Entity> surface = AZStd::make_unique<AZ::Entity>();
+                AZ::Vector3 worldPos(halfWorldSize, halfWorldSize, 10.0f);
+
+                auto transform = surface->CreateComponent<AzFramework::TransformComponent>();
+                transform->SetLocalTM(AZ::Transform::CreateTranslation(worldPos));
+                transform->SetWorldTM(AZ::Transform::CreateTranslation(worldPos));
```
Comment (invertednormal) [inline Gems/GradientSignal/Code/Tests/GradientSignalBenchmarks.cpp:200]: nit: I'm pretty sure when you set the world tm, it also sets the local, so the above call is probably unnecessary.

File: Gems/SurfaceData/Code/Tests/SurfaceDataTest.cpp [hunk 2]
```diff
@@ -167,19 +167,20 @@ class MockSurfaceProvider
...
+            AZStd::span<SurfaceData::SurfaceTagWeights> weights) const override
         {
-            surfacePointList.ModifySurfaceWeights(
-                AZ::EntityId(),
-                [this](const AZ::Vector3& position, SurfaceData::SurfaceTagWeights& weights)
-                {
-                    auto surfacePoints = m_GetSurfacePoints.find(AZStd::make_pair(position.GetX(), position.GetY()));
+            for (size_t index = 0; index < positions.size(); index++)
+            {
+                auto surfacePoints = m_GetSurfacePoints.find(AZStd::make_pair(positions[index].GetX(), positions[index].GetY()));
```
Comment (invertednormal) [inline Gems/SurfaceData/Code/Tests/SurfaceDataTest.cpp:177]: nit: `m_GetSurfacePoints` isn't correct naming convention (although the problem existed before this review).

**Review by invertednormal — APPROVED** [review]
Looks good, just a couple nits but nothing blocking.

File: Gems/GradientSignal/Code/Tests/GradientSignalBenchmarks.cpp (hunk 1 above)
Comment (mbalfour-amzn) [reply Gems/GradientSignal/Code/Tests/GradientSignalBenchmarks.cpp:199]: Removed all extra SetLocalTM calls.

File: Gems/SurfaceData/Code/Tests/SurfaceDataTest.cpp (hunk 2 above)
Comment (mbalfour-amzn) [reply Gems/SurfaceData/Code/Tests/SurfaceDataTest.cpp]: Renamed to m_surfacePoints.

**Review by amzn-sj — APPROVED** [review]

**Pull request merged**